tests/async/test_browsercontext_add_cookies.py

```diff
@@ -0,0 +1,387 @@
+    await context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "password", "value": "123456"}]
+    )
+    assert await page.evaluate("() => document.cookie") == "password=123456"
+
+
+async def test_should_roundtrip_cookie(context, page, server):
+    await page.goto(server.EMPTY_PAGE)
+    # @see https://en.wikipedia.org/wiki/Year_2038_problem
+    date = int(datetime.datetime(2038, 1, 1).timestamp() * 1000)
+    document_cookie = await page.evaluate(
+        """timestamp => {
+    const date = new Date(timestamp);
+    document.cookie = `username=John Doe;expires=${date.toUTCString()}`;
+    return document.cookie;
+  }""",
+        date,
+    )
+    assert document_cookie == "username=John Doe"
+    cookies = await context.cookies()
+    await context.clearCookies()
+    assert await context.cookies() == []
+    await context.addCookies(cookies)
+    assert await context.cookies() == cookies
+
+
+async def test_should_send_cookie_header(server, context):
+    cookie = []
+
+    def handler(request):
+        cookie.extend(request.requestHeaders.getRawHeaders("cookie"))
+        request.finish()
+
+    server.set_route("/empty.html", handler)
+    await context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "cookie", "value": "value"}]
+    )
+    page = await context.newPage()
+    await page.goto(server.EMPTY_PAGE)
+    assert cookie == ["cookie=value"]
+
+
+async def test_should_isolate_cookies_in_browser_contexts(context, server, browser):
+    another_context = await browser.newContext()
+    await context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "isolatecookie", "value": "page1value"}]
+    )
+    await another_context.addCookies(
+        [{"url": server.EMPTY_PAGE, "name": "isolatecookie", "value": "page2value"}]
+    )
+
+    cookies_1 = await context.cookies()
+    cookies_2 = await another_context.cookies()
+    assert len(cookies_1) == 1
+    assert len(cookies_2) == 1
+    assert cookies_1[0]["name"] == "isolatecookie"
+    assert cookies_1[0]["value"] == "page1value"
+    assert cookies_2[0]["name"] == "isolatecookie"
+    assert cookies_2[0]["value"] == "page2value"
+    await another_context.close()
+
+
+async def test_should_isolate_session_cookies(context, server, browser):
+    server.set_route(
+        "/setcookie.html",
+        lambda r: (r.setHeader("Set-Cookie", "session=value"), r.finish(),),
+    )
+
+    page_1 = await context.newPage()
+    await page_1.goto(server.PREFIX + "/setcookie.html")
+    ##
+    page_2 = await context.newPage()
+    await page_2.goto(server.EMPTY_PAGE)
+    cookies_2 = await context.cookies()
+    assert len(cookies_2) == 1
+    assert ",".join(list(map(lambda c: c["value"], cookies_2))) == "value"
+    ##
+    context_b = await browser.newContext()
+    page_3 = await context_b.newPage()
+    await page_3.goto(server.EMPTY_PAGE)
+    cookies_3 = await context_b.cookies()
+    assert cookies_3 == []
+    await context_b.close()
```

Comment:
The javascript version used scoping; I've manually appended subscripts to (hopefully) achieve the same thing. The few tests with `##` should be double checked by someone else to ensure the logic is preserved. 👀